### PR TITLE
Implement Process Crosslink From 0.6

### PR DESCRIPTION
**Pull request by terencechain**
This is part of #2307 

In this PR we implement new cross link processing from 0.6

Changed:
- [x] Updated `WinningCrosslink` to return winning crosslink's attesting indices
- [x] Implemented `ProcessCrosslink`
- [x] Test `ProcessCrosslink`

### CODE REVIEW

File: beacon-chain/core/epoch/epoch_processing.go [hunk 1]
```diff
@@ -556,6 +556,45 @@ func MatchAttestations(state *pb.BeaconState, epoch uint64) (*MatchedAttestation
 	}, nil
 }
 
+// ProcessCrosslink processes crosslink and finds the crosslink
+// with enough state to make it canonical in state.
+//
+// Spec pseudocode definition:
+//	def process_crosslinks(state: BeaconState) -> None:
+//    state.previous_crosslinks = [c for c in state.current_crosslinks]
```
Comment (prestonvanloon) [inline beacon-chain/core/epoch/epoch_processing.go]: Formatting

File: beacon-chain/core/epoch/epoch_processing.go [hunk 2]
```diff
@@ -556,6 +556,45 @@ func MatchAttestations(state *pb.BeaconState, epoch uint64) (*MatchedAttestation
...
+			return nil, err
+		}
+		for _, committee := range committees {
+			crosslink, indices, err := WinningCrosslink(state, committee.Shard, epoch)
+			if err != nil {
+				return nil, err
+			}
+			attestedBalance := TotalBalance(state, indices)
+			totalBalance := TotalBalance(state, committee.Committee)
+			if 3*attestedBalance >= 2*totalBalance {
```
Comment (prestonvanloon) [inline beacon-chain/core/epoch/epoch_processing.go]: What is the significance of 3?
What is the significance of 2?

These appear to be magic numbers. Please consider a named constant or inline comment to explain why 2, 3.

Comment (terencechain) [reply beacon-chain/core/epoch/epoch_processing.go]: Sounds, adding a comment. 
I tried to introduce 2/3 to `config.go` but it'll introduce `float64`. `2 * x > 3 * y` feels better in this case

Comment (prestonvanloon) [reply beacon-chain/core/epoch/epoch_processing.go]: sg

File: beacon-chain/core/epoch/epoch_processing.go [hunk 3]
```diff
@@ -725,31 +768,38 @@ func WinningCrosslink(state *pb.BeaconState, shard uint64, epoch uint64) (*pb.Cr
...
 		}
 		if attestingBalance > winnerBalance {
 			winnerCrosslink = c
 		}
 	}
 
-	return winnerCrosslink, nil
+	crosslinkIndices, err := UnslashedAttestingIndices(state, attsForCrosslink(state, winnerCrosslink, shardAtts))
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not get crosslink indices")
```
Comment (prestonvanloon) [inline beacon-chain/core/epoch/epoch_processing.go]:
```suggestion
		return nil, nil, errors.New("could not get crosslink indices")
```

File: beacon-chain/core/epoch/epoch_processing.go (hunk 1 above)
Comment (prestonvanloon) [reply beacon-chain/core/epoch/epoch_processing.go]: Still formatting

image

**Pull request merged**